Assistive Technology not found: com.sun.java.accessibility.AccessBridge

When QZ Tray is starting on a PC with Java 1.8 assistive technologies enabled, the following error occurs.
* Configuration file `%USERPROFILE%\.accessibility.properties`
   ```properties
   assistive_technologies=com.sun.java.accessibility.AccessBridge
   screen_magnifier_present=true
   ```

Workaround:
* Delete or comment the line containing the missing class file.
   ```properties
   #assistive_technologies=com.sun.java.accessibility.AccessBridge
   screen_magnifier_present=true
   ```
* Credit goes to https://stackoverflow.com/a/74178775/3196753

```
[INFO] 2024-02-14T19:32:51,524 @ qz.App:70
        Starting QZ Tray 2.2.2+1
Exception in thread "main" java.awt.AWTError: Assistive Technology not found: com.sun.java.accessibility.AccessBridge
        at java.desktop/java.awt.Toolkit.newAWTError(Unknown Source)
        at java.desktop/java.awt.Toolkit.fallbackToLoadClassForAT(Unknown Source)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(Unknown Source)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(Unknown Source)
        at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(Unknown Source)
        at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(Unknown Source)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(Unknown Source)
        at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)
        at java.base/java.util.stream.ReferencePipeline.forEach(Unknown Source)
        at java.desktop/java.awt.Toolkit.loadAssistiveTechnologies(Unknown Source)
        at java.desktop/java.awt.Toolkit.getDefaultToolkit(Unknown Source)
        at java.desktop/java.awt.EventQueue.invokeAndWait(Unknown Source)
        at java.desktop/javax.swing.SwingUtilities.invokeAndWait(Unknown Source)
        at qz.ws.PrintSocketServer.runServer(PrintSocketServer.java:60)
        at qz.App.main(App.java:72)
Caused by: java.lang.ClassNotFoundException: com.sun.java.accessibility.AccessBridge
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(Unknown Source)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(Unknown Source)
        at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Unknown Source)
        ... 16 more
[INFO] 2024-02-14T19:33:51,305 @ qz.installer.certificate.CertificateManager:145
        SSL certificate is still valid for 824 more days: 2026-05-18T16:31Z.  We'll make a new one automatically when needed.Here are some more logs.
```

--- a/src/qz/build/JLink.java
+++ b/src/qz/build/JLink.java
@@ -233,7 +233,7 @@ public class JLink {
         }
         switch(targetPlatform) {
             case WINDOWS:
-                // Java accessibility bridge dependency, see https://github.com/adoptium/adoptium-support/issues/1234
+                // Java accessibility bridge dependency, see https://github.com/qzind/tray/issues/1234
                 depList.add("jdk.accessibility");
             default:
                 // "jar:" URLs create transient zipfs dependency, see https://stackoverflow.com/a/57846672/3196753

--- a/src/qz/build/JLink.java
+++ b/src/qz/build/JLink.java
@@ -26,9 +26,7 @@ import java.io.File;
 import java.io.FileInputStream;
 import java.io.IOException;
 import java.nio.file.*;
-import java.util.HashMap;
-import java.util.LinkedHashSet;
-import java.util.Properties;
+import java.util.*;
 
 public class JLink {
     private static final Logger log = LogManager.getLogger(JLink.class);
@@ -233,12 +231,16 @@ public class JLink {
                 depList.add(item);
             }
         }
-        // the jdk.accessibility package is used if JAB is enabled; JDeps misses this, leading to a missing class runtime error. see #1234
-        if (targetPlatform == Platform.WINDOWS) depList.add("jdk.accessibility");
-        // "jar:" URLs create transient zipfs dependency, see https://stackoverflow.com/a/57846672/3196753
-        depList.add("jdk.zipfs");
-        // fix for https://github.com/qzind/tray/issues/894 solution from https://github.com/adoptium/adoptium-support/issues/397
-        depList.add("jdk.crypto.ec");
+        switch(targetPlatform) {
+            case WINDOWS:
+                // Java accessibility bridge dependency, see https://github.com/adoptium/adoptium-support/issues/1234
+                depList.add("jdk.accessibility");
+            default:
+                // "jar:" URLs create transient zipfs dependency, see https://stackoverflow.com/a/57846672/3196753
+                depList.add("jdk.zipfs");
+                // fix for https://github.com/qzind/tray/issues/894 solution from https://github.com/adoptium/adoptium-support/issues/397
+                depList.add("jdk.crypto.ec");
+        }
         return this;
     }
 
@@ -281,17 +283,22 @@ public class JLink {
             log.info("Successfully deployed a jre to {}", outPath);
 
             // Remove all but java/javaw
-            String[] keepFiles;
+            List<String> keepFiles = new ArrayList<>();
+            //String[] keepFiles;
             String keepExt;
             switch(targetPlatform) {
                 case WINDOWS:
-                    // Jabswitch is a tool for enabling and disabling accessibility support on windows
-                    keepFiles = new String[]{ "java.exe", "javaw.exe", "jabswitch.exe"};
+                    keepFiles.add("java.exe");
+                    keepFiles.add("javaw.exe");
+                    if(depList.contains("jdk.accessibility")) {
+                        // Java accessibility bridge switching tool
+                        keepFiles.add("jabswitch.exe");
+                    }
                     // Windows stores ".dll" files in bin
                     keepExt = ".dll";
                     break;
                 default:
-                    keepFiles = new String[]{ "java" };
+                    keepFiles.add("java");
                     keepExt = null;
             }
 

--- a/src/qz/build/JLink.java
+++ b/src/qz/build/JLink.java
@@ -233,6 +233,8 @@ public class JLink {
                 depList.add(item);
             }
         }
+        // the jdk.accessibility package is used if JAB is enabled; JDeps misses this, leading to a missing class runtime error. see #1234
+        if (targetPlatform == Platform.WINDOWS) depList.add("jdk.accessibility");
         // "jar:" URLs create transient zipfs dependency, see https://stackoverflow.com/a/57846672/3196753
         depList.add("jdk.zipfs");
         // fix for https://github.com/qzind/tray/issues/894 solution from https://github.com/adoptium/adoptium-support/issues/397
@@ -283,7 +285,8 @@ public class JLink {
             String keepExt;
             switch(targetPlatform) {
                 case WINDOWS:
-                    keepFiles = new String[]{ "java.exe", "javaw.exe" };
+                    // Jabswitch is a tool for enabling and disabling accessibility support on windows
+                    keepFiles = new String[]{ "java.exe", "javaw.exe", "jabswitch.exe"};
                     // Windows stores ".dll" files in bin
                     keepExt = ".dll";
                     break;

--- a/src/qz/ui/DetailsDialog.java
+++ b/src/qz/ui/DetailsDialog.java
@@ -31,12 +31,16 @@ public class DetailsDialog extends JPanel {
 
         requestTable = new RequestTable(iconCache);
         reqScrollPane = new JScrollPane(requestTable);
+        requestTable.getAccessibleContext().setAccessibleName("Request Details");
+        requestTable.getAccessibleContext().setAccessibleDescription("This table provides details about this request.");
 
         certLabel = new JLabel("Certificate");
         certLabel.setAlignmentX(CENTER_ALIGNMENT);
 
         certTable = new CertificateTable(iconCache);
         certScrollPane = new JScrollPane(certTable);
+        certTable.getAccessibleContext().setAccessibleName("Certificate Details");
+        certTable.getAccessibleContext().setAccessibleDescription("This table provides details about the certificate used in this request.");
 
         add(requestLabel);
         add(reqScrollPane);

--- a/src/qz/ui/DetailsDialog.java
+++ b/src/qz/ui/DetailsDialog.java
@@ -31,16 +31,18 @@ public class DetailsDialog extends JPanel {
 
         requestTable = new RequestTable(iconCache);
         reqScrollPane = new JScrollPane(requestTable);
-        requestTable.getAccessibleContext().setAccessibleName("Request Details");
-        requestTable.getAccessibleContext().setAccessibleDescription("This table provides details about this request.");
+        requestTable.getAccessibleContext().setAccessibleName(requestLabel.getText() + " Details");
+        requestTable.getAccessibleContext().setAccessibleDescription("Signing details about this request.");
+        requestLabel.setLabelFor(requestTable);
 
         certLabel = new JLabel("Certificate");
         certLabel.setAlignmentX(CENTER_ALIGNMENT);
 
         certTable = new CertificateTable(iconCache);
         certScrollPane = new JScrollPane(certTable);
-        certTable.getAccessibleContext().setAccessibleName("Certificate Details");
-        certTable.getAccessibleContext().setAccessibleDescription("This table provides details about the certificate used in this request.");
+        certTable.getAccessibleContext().setAccessibleName(certLabel.getText() + " Details");
+        certTable.getAccessibleContext().setAccessibleDescription("Certificate details about this request.");
+        certLabel.setLabelFor(certTable);
 
         add(requestLabel);
         add(reqScrollPane);

--- a/src/qz/ui/component/EmLabel.java
+++ b/src/qz/ui/component/EmLabel.java
@@ -7,25 +7,25 @@ import java.util.HashMap;
 import java.util.Map;
 
 /**
- * Create a label at the multiplier of it's normal size, similar to CSS's "em" tag
+ * Create a label at the multiplier of its normal size, similar to CSS's "em" tag
  */
 public class EmLabel extends JLabel {
-    public EmLabel() {}
-    public EmLabel(String text) {
-        super(text);
-    }
     public EmLabel(String text, float multiplier) {
         this(text, multiplier, true);
     }
     public EmLabel(String text, float multiplier, boolean underline) {
         super(text);
-        Font template = getFont().deriveFont(multiplier * getFont().getSize());
+        stylizeComponent(this, multiplier, underline);
+    }
+
+    public static void stylizeComponent(Component j, float multiplier, boolean underline) {
+        Font template = j.getFont().deriveFont(multiplier * j.getFont().getSize());
         if (!underline) {
             Map<TextAttribute,Object> attributes = new HashMap<>(template.getAttributes());
             attributes.remove(TextAttribute.UNDERLINE);
-            setFont(template.deriveFont(attributes));
+            j.setFont(template.deriveFont(attributes));
         } else {
-            setFont(template);
+            j.setFont(template);
         }
     }
 }

--- a/src/qz/ui/component/LinkLabel.java
+++ b/src/qz/ui/component/LinkLabel.java
@@ -8,11 +8,9 @@ import qz.utils.ShellUtilities;
 
 import javax.accessibility.AccessibleContext;
 import javax.accessibility.AccessibleRole;
+import javax.swing.*;
 import java.awt.*;
-import java.awt.event.ActionEvent;
-import java.awt.event.ActionListener;
-import java.awt.event.MouseEvent;
-import java.awt.event.MouseListener;
+import java.awt.event.*;
 import java.awt.font.TextAttribute;
 import java.io.File;
 import java.net.MalformedURLException;
@@ -72,7 +70,6 @@ public class LinkLabel extends EmLabel implements Themeable {
         addActionListener(ae -> ShellUtilities.browseDirectory(filePath.isDirectory()? filePath.getPath():filePath.getParent()));
     }
 
-
     private void initialize() {
         Map<TextAttribute,Object> attributes = new HashMap<>(getFont().getAttributes());
         attributes.put(TextAttribute.UNDERLINE, TextAttribute.UNDERLINE_ON);
@@ -80,6 +77,17 @@ public class LinkLabel extends EmLabel implements Themeable {
         setFocusable(true);
 
         actionListeners = new ArrayList<>();
+
+        this.getInputMap(JComponent.WHEN_FOCUSED).put(KeyStroke.getKeyStroke("released SPACE"), "pressed");
+        this.getInputMap(JComponent.WHEN_FOCUSED).put(KeyStroke.getKeyStroke("pressed ENTER"), "pressed");
+        this.getActionMap().put("pressed", new AbstractAction() {
+            @Override
+            public void actionPerformed(ActionEvent e) {
+                for(ActionListener actionListener : actionListeners) {
+                    actionListener.actionPerformed(new ActionEvent(e.getSource(), e.getID(), "keyClicked"));
+                }
+            }
+        });
 
         addMouseListener(new MouseListener() {
             @Override

--- a/src/qz/ui/component/LinkLabel.java
+++ b/src/qz/ui/component/LinkLabel.java
@@ -6,6 +6,8 @@ import qz.common.Constants;
 import qz.ui.Themeable;
 import qz.utils.ShellUtilities;
 
+import javax.accessibility.AccessibleContext;
+import javax.accessibility.AccessibleRole;
 import javax.swing.*;
 import java.awt.*;
 import java.awt.font.TextAttribute;
@@ -80,5 +82,18 @@ public class LinkLabel extends JButton implements Themeable {
         setBorder(null);
         setOpaque(false);
         setCursor(Cursor.getPredefinedCursor(Cursor.HAND_CURSOR));
+    }
+
+    public AccessibleContext getAccessibleContext() {
+        if (accessibleContext == null) {
+            accessibleContext = new AccessibleLinkLabel();
+        }
+        return accessibleContext;
+    }
+
+    protected class AccessibleLinkLabel extends AccessibleJButton {
+        public AccessibleRole getAccessibleRole() {
+            return AccessibleRole.HYPERLINK;
+        }
     }
 }

--- a/src/qz/ui/component/LinkLabel.java
+++ b/src/qz/ui/component/LinkLabel.java
@@ -6,6 +6,8 @@ import qz.common.Constants;
 import qz.ui.Themeable;
 import qz.utils.ShellUtilities;
 
+import javax.accessibility.AccessibleContext;
+import javax.accessibility.AccessibleRole;
 import java.awt.*;
 import java.awt.event.ActionEvent;
 import java.awt.event.ActionListener;
@@ -75,6 +77,7 @@ public class LinkLabel extends EmLabel implements Themeable {
         Map<TextAttribute,Object> attributes = new HashMap<>(getFont().getAttributes());
         attributes.put(TextAttribute.UNDERLINE, TextAttribute.UNDERLINE_ON);
         setFont(getFont().deriveFont(attributes));
+        setFocusable(true);
 
         actionListeners = new ArrayList<>();
 
@@ -121,4 +124,18 @@ public class LinkLabel extends EmLabel implements Themeable {
         actionListeners.remove(action);
     }
 
+    @Override
+    public AccessibleContext getAccessibleContext() {
+        if (accessibleContext == null) {
+            accessibleContext = new AccessibleLinkLabel();
+        }
+        return accessibleContext;
+    }
+
+    protected class AccessibleLinkLabel extends AccessibleJLabel {
+        @Override
+        public AccessibleRole getAccessibleRole() {
+            return AccessibleRole.HYPERLINK;
+        }
+    }
 }

--- a/src/qz/ui/component/LinkLabel.java
+++ b/src/qz/ui/component/LinkLabel.java
@@ -79,7 +79,6 @@ public class LinkLabel extends JButton implements Themeable {
         setBorderPainted(false);
         setBorder(null);
         setOpaque(false);
-        setFocusable(true);
         setCursor(Cursor.getPredefinedCursor(Cursor.HAND_CURSOR));
     }
 }

--- a/src/qz/ui/component/LinkLabel.java
+++ b/src/qz/ui/component/LinkLabel.java
@@ -6,27 +6,25 @@ import qz.common.Constants;
 import qz.ui.Themeable;
 import qz.utils.ShellUtilities;
 
-import javax.accessibility.AccessibleContext;
-import javax.accessibility.AccessibleRole;
 import javax.swing.*;
 import java.awt.*;
-import java.awt.event.*;
 import java.awt.font.TextAttribute;
 import java.io.File;
 import java.net.MalformedURLException;
 import java.net.URL;
-import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.Map;
 
 /**
+ * Creates a JButton which visually appears as a clickable link
+ *
+ * TODO: Rename this class.  Since switching from JLabel to a JButton, this class now has a misleading name.
+ *
  * Created by Tres on 2/19/2015.
  */
-public class LinkLabel extends EmLabel implements Themeable {
+public class LinkLabel extends JButton implements Themeable {
 
     private static final Logger log = LogManager.getLogger(LinkLabel.class);
-
-    private ArrayList<ActionListener> actionListeners;
 
     public LinkLabel() {
         super();
@@ -39,7 +37,8 @@ public class LinkLabel extends EmLabel implements Themeable {
     }
 
     public LinkLabel(String text, float multiplier, boolean underline) {
-        super(text, multiplier, underline);
+        super(text);
+        EmLabel.stylizeComponent(this, multiplier, underline);
         initialize();
     }
 
@@ -53,15 +52,12 @@ public class LinkLabel extends EmLabel implements Themeable {
     }
 
     public void setLinkLocation(final URL location) {
-        addActionListener(new ActionListener() {
-            @Override
-            public void actionPerformed(ActionEvent ae) {
-                try {
-                    Desktop.getDesktop().browse(location.toURI());
-                }
-                catch(Exception e) {
-                    log.error("", e);
-                }
+        addActionListener(ae -> {
+            try {
+                Desktop.getDesktop().browse(location.toURI());
+            }
+            catch(Exception e) {
+                log.error("", e);
             }
         });
     }
@@ -74,76 +70,16 @@ public class LinkLabel extends EmLabel implements Themeable {
         Map<TextAttribute,Object> attributes = new HashMap<>(getFont().getAttributes());
         attributes.put(TextAttribute.UNDERLINE, TextAttribute.UNDERLINE_ON);
         setFont(getFont().deriveFont(attributes));
-        setFocusable(true);
-
-        actionListeners = new ArrayList<>();
-
-        this.getInputMap(JComponent.WHEN_FOCUSED).put(KeyStroke.getKeyStroke("released SPACE"), "pressed");
-        this.getInputMap(JComponent.WHEN_FOCUSED).put(KeyStroke.getKeyStroke("pressed ENTER"), "pressed");
-        this.getActionMap().put("pressed", new AbstractAction() {
-            @Override
-            public void actionPerformed(ActionEvent e) {
-                for(ActionListener actionListener : actionListeners) {
-                    actionListener.actionPerformed(new ActionEvent(e.getSource(), e.getID(), "keyClicked"));
-                }
-            }
-        });
-
-        addMouseListener(new MouseListener() {
-            @Override
-            public void mouseClicked(MouseEvent e) {
-                for(ActionListener actionListener : actionListeners) {
-                    actionListener.actionPerformed(new ActionEvent(e.getSource(), e.getID(), "mouseClicked"));
-                }
-            }
-
-            @Override
-            public void mousePressed(MouseEvent e) {}
-
-            @Override
-            public void mouseReleased(MouseEvent e) {}
-
-            @Override
-            public void mouseEntered(MouseEvent e) {
-                setCursor(Cursor.getPredefinedCursor(Cursor.HAND_CURSOR));
-            }
-
-            @Override
-            public void mouseExited(MouseEvent e) {
-                setCursor(Cursor.getDefaultCursor());
-            }
-        });
-
         refresh();
     }
 
     @Override
     public void refresh() {
         setForeground(Constants.TRUSTED_COLOR);
-    }
-
-    public void addActionListener(ActionListener action) {
-        if (!actionListeners.contains(action)) {
-            actionListeners.add(action);
-        }
-    }
-
-    public void removeActionListener(ActionListener action) {
-        actionListeners.remove(action);
-    }
-
-    @Override
-    public AccessibleContext getAccessibleContext() {
-        if (accessibleContext == null) {
-            accessibleContext = new AccessibleLinkLabel();
-        }
-        return accessibleContext;
-    }
-
-    protected class AccessibleLinkLabel extends AccessibleJLabel {
-        @Override
-        public AccessibleRole getAccessibleRole() {
-            return AccessibleRole.HYPERLINK;
-        }
+        setBorderPainted(false);
+        setBorder(null);
+        setOpaque(false);
+        setFocusable(true);
+        setCursor(Cursor.getPredefinedCursor(Cursor.HAND_CURSOR));
     }
 }

